### ISSUE TITLE
Make eslint work with requireType

### ISSUE THREAD

**Issue by maribethb** [google/blockly#5161]
**Is your feature request related to a problem? Please describe.**

Types only used in JsDoc (and imported as `const Foo = goog.requireType('Blockly.Foo');`) are "unused" in code and therefore will trip the "no-unused-var" rule in eslint. 

**Describe the solution you'd like**

There is a Google-owned library that claims to implement the Google style guide rules and also understand Closure: https://github.com/google/eslint-closure

This plugin also comes with a config that enables many rules that it claims follow Google style. However, many of these rules are more strict than Google style and do not match the style we're using in blockly-samples. They conflict with the rules in eslint-config-google (see #5160). Therefore, we can't just use the config included with this plugin.

In order to get just the bit of this plugin that marks JsDoc symbols as "used," we need to include it in the `plugins` section of the config and then we need to enable the `closure/jsdoc` rule. I am still working out the exact settings necessary. But it's during the processing of this rule that the variables are marked as used.

I'm not sure yet how this works with the jsdoc plugin (see #2347). I will be testing that out.

**Describe alternatives you've considered**

- Create our own plugin / rule that understands Closure and won't yell about no-unused-vars for requireType, but doesn't do extra style rules on top
    - This involves parsing the code into an AST and understanding even more about eslint than I do now...
- Disable eslint manually for every `requireType` line
    - This is the interim solution but it's manual toil and ugly
- Disable the "no-unused-vars" rule completely
    - Not a great solution because this could catch legitimate bugs e.g. incomplete refactoring

**Comment by maribethb** [comment]
So, after more testing, I realized the eslint-closure plugin won't actually work, at least not fully. It can understand *some* type comments, but it doesn't handle all of them, such as ones that are used in typecasts. I don't think there is any configuration that can help with this, as even in the demo page, it would complain about typecasts type variables not being used.

The jsdoc plugin that is recommended by eslint also does not solve this problem. It actually has a rule that does something very similar to the closure plugin, `jsdoc/no-undefined-types` that 1) makes sure all types are required/imported into the file and 2) marks types as used so that `no-unused-types` doesn't complain about them. However, it also does not check type variables used in typecasts only. It at least makes sense for this one since type casts are purely a Closure construct and not a JsDoc thing at all. But this rule causes additional problems as well because it complains that certain tags are undefined types, such as `@suppress {extraRequire}` it complains that `extraRequire` is not defined. So we'd have to list out every Closure suppression in order to convince this plugin to enforce this rule correctly.

I spent some time looking at parsing JsDoc into AST, but that story is even more complicated than I originally thought. eslint used to support JsDoc out of the box, and they had their own parser for JsDoc.  However, they deprecated not only eslint direct support for JsDoc, but also their parser project Doctrine. There are a handful of other parsing projects in use but none of them have great documentation stories and afaict they don't have interactive AST explorers that would simplify development. It seems like if we ever did want to create our own rule, a good starting place would be dissecting the eslint-plugin-jsdoc/no-undefined-types rule, but for now it's not a good enough use of our time.

**Comment by cpcallen** [comment]
I think the path I'd try to explore is modifying eslint-closure to look for `@type` tags too.  Have you looked to see if [Doctrine](https://github.com/eslint/doctrine) will parse them correctly, and (if so) if eslint-closure can be modified to get them parsed?

The fact that Doctrine is deprecated is neither here nor there.  The code is out there and if it works, it works.  We don't need an infinitely future-proof solution: since we will probably move to TypeScript before Closure Compiler starts making changes to its input language that will be incompatible with Doctrine.

Of course if Doctrine can't deal with `@type` at the moment (and can't be easily modified to either) then we may be stuck with eslint disables.

**Comment by maribethb** [comment]
My problem with Doctrine is that the extent of the documentation is located on that README. They used to have a demo page with an interactive AST explorer but that no longer exists. So if we were going to modify eslint-closure then we could probably figure it out, but if we were going to modify the jsdoc plugin (or copy it and modify it) then we'd be better off using whatever parser they're using.

My problem with eslint-closure is that it does a bunch of stuff I don't want it to do :/ It claims to follow the Google style guide but it also implements a lot of rules I think are stricter than the Google style guide. So I got around this by only using the jsdoc rule, but then that rule is still inferior to the eslint-jsdoc plugin and we'd have two rules trying to do the same thing, or at least overlapping. My goal with the custom rule would be that it doesn't actually ever report errors, it only marks variables as used and we rely on other rules to do actual enforcement.

Also, the original author of eslint-closure has left Google and I'm not sure that the package is still being maintained in any way. It's only ever been published once so it might be the case that even if a PR were accepted it would not be published without our involvement. And if we were going to do all that, I might rather not just stick this into a plugin we don't want the rest of. Whereas even if we can't get our changes merged into the jsdoc plugin, it's at least under active development and we could probably ask them questions or possibly get some parts merged in (I think the way they deal with `@suppress` tags is a bug, imo, or at least a reasonable feature request since they support Closure mode)

I think Doctrine does understand `@type` annotations because it's the same annotation used e.g. to declare the type of a property. But doctrine only handles JsDoc strings, so you need to parse the raw code correctly, and that's where I kind of lose sight of how to do that correctly. There used to be methods built in directly to eslint, that have now been replaced by another package, that can get a comment associated with a particular node. and I'm not sure how they deal with the typecast comments or even if we need to care about that, or if we can just grab every single comment in a file. If you play with the Esprima parser in AST Explorer, it doesn't show comments at all. So that's kind of where I got stuck.

So, to recap, I personally would be more in favor of modifying the eslint-jsdoc plugin than the eslint-closure one, but ultimately for where I'm at in my knowledge, it wouldn't be a good use of my time at this moment.

**Comment by maribethb** [comment]
Obsolete due to move to ts/esm
